jerry-core/ecma/base/ecma-helpers-conversion.c

```diff
@@ -917,227 +919,512 @@ ecma_number_to_int32 (ecma_number_t num) /**< ecma-number */
+  POWER_CACHE_ITEM (0xf209787bb47d6b85ULL,   348,  124),
+  POWER_CACHE_ITEM (0xb454e4a179dd1877ULL,   375,  132),
+  POWER_CACHE_ITEM (0x865b86925b9bc5c2ULL,   402,  140),
+  POWER_CACHE_ITEM (0xc83553c5c8965d3dULL,   428,  148),
+  POWER_CACHE_ITEM (0x952ab45cfa97a0b3ULL,   455,  156),
+  POWER_CACHE_ITEM (0xde469fbd99a05fe3ULL,   481,  164),
+  POWER_CACHE_ITEM (0xa59bc234db398c25ULL,   508,  172),
+  POWER_CACHE_ITEM (0xf6c69a72a3989f5cULL,   534,  180),
+  POWER_CACHE_ITEM (0xb7dcbf5354e9beceULL,   561,  188),
+  POWER_CACHE_ITEM (0x88fcf317f22241e2ULL,   588,  196),
+  POWER_CACHE_ITEM (0xcc20ce9bd35c78a5ULL,   614,  204),
+  POWER_CACHE_ITEM (0x98165af37b2153dfULL,   641,  212),
+  POWER_CACHE_ITEM (0xe2a0b5dc971f303aULL,   667,  220),
+  POWER_CACHE_ITEM (0xa8d9d1535ce3b396ULL,   694,  228),
+  POWER_CACHE_ITEM (0xfb9b7cd9a4a7443cULL,   720,  236),
+  POWER_CACHE_ITEM (0xbb764c4ca7a44410ULL,   747,  244),
+  POWER_CACHE_ITEM (0x8bab8eefb6409c1aULL,   774,  252),
+  POWER_CACHE_ITEM (0xd01fef10a657842cULL,   800,  260),
+  POWER_CACHE_ITEM (0x9b10a4e5e9913129ULL,   827,  268),
+  POWER_CACHE_ITEM (0xe7109bfba19c0c9dULL,   853,  276),
+  POWER_CACHE_ITEM (0xac2820d9623bf429ULL,   880,  284),
+  POWER_CACHE_ITEM (0x80444b5e7aa7cf85ULL,   907,  292),
+  POWER_CACHE_ITEM (0xbf21e44003acdd2dULL,   933,  300),
+  POWER_CACHE_ITEM (0x8e679c2f5e44ff8fULL,   960,  308),
+  POWER_CACHE_ITEM (0xd433179d9c8cb841ULL,   986,  316),
+  POWER_CACHE_ITEM (0x9e19db92b4e31ba9ULL,  1013,  324),
+  POWER_CACHE_ITEM (0xeb96bf6ebadf77d9ULL,  1039,  332),
+  POWER_CACHE_ITEM (0xaf87023b9bf0ee6bULL,  1066,  340)
+};
 
-        decimal_exp--;
-      }
-    }
-  }
+/**
+ * Get power from the chache
+ *
+ * Note:
+ *      Section 4. (ISBN: 978-1-4503-0019-3)
+ *
+ * @return chached power
+ */
+static int
+ecma_find_power_in_cache (int exp, /**< exponent */
+                          encoded_float_t *fp_value) /**< float value */
+{
+  int k = (int) ceil ((exp + GRISU3_ENCODED_FLOAT_FRACT_SIZE - 1) * GRISU3_1_LOG2_10); /* k_computation */
+  int i = (k - GRISU3_MIN_CACHED_EXP - 1) / GRISU3_CACHED_EXP_STEP + 1;
+  fp_value->fract = power_cache[i].fract;
+  fp_value->exp = power_cache[i].bin_exp;
+  return power_cache[i].dec_exp;
+} /* ecma_find_power_in_cache */
 
-  *out_decimal_exp_p = decimal_exp;
-} /* ecma_number_helper_binary_to_decimal */
+/**
+ * Subtract two floating point values
+ *
+ * Note:
+ *      Figure 2/a (ISBN: 978-1-4503-0019-3)
+ *
+ * @return result of subtraction
+ */
+static encoded_float_t
+ecma_subtract_encoded_fp (encoded_float_t fp_value_1, /**< first float value */
+                          encoded_float_t fp_value_2) /**< second float value */
+{
+  encoded_float_t result;
 
-#endif /* CONFIG_ECMA_NUMBER_TYPE == CONFIG_ECMA_NUMBER_FLOAT64 */
+  result.fract = fp_value_1.fract - fp_value_2.fract;
+  result.exp = fp_value_1.exp;
+
+  JERRY_ASSERT (fp_value_1.exp == fp_value_2.exp && fp_value_1.fract >= fp_value_2.fract);
```

Comment:
Since this assert checks the input, it should better be the first statement in the function.